internal/web/validate.go

```diff
@@ -150,114 +322,48 @@ func runValidator(validator, repopath, commit string, gcl *ginclient.Client) (in
-	resdir := filepath.Join(srvcfg.Dir.Result, validator, repopath, srvcfg.Label.ResultsFolder)
 	err = os.MkdirAll(resdir, os.ModePerm)
 	if err != nil {
-		log.ShowWrite("[Error] creating '%s' results folder: %s", repopath, err.Error())
+		log.ShowWrite("[Error] creating %q results folder: %s", valroot, err.Error())
 		return http.StatusInternalServerError, fmt.Errorf("failed to generate results")
 	}
 
-	// Use validation config file if available
-	valroot := filepath.Join(tmpdir, repo)
-	var validateNifti bool
-
-	cfgpath := filepath.Join(tmpdir, repo, srvcfg.Label.ValidationConfigFile)
-	log.ShowWrite("[Info] looking for config file at '%s'", cfgpath)
-	if fi, err := os.Stat(cfgpath); err == nil && !fi.IsDir() {
-		valcfg, err := handleValidationConfig(cfgpath)
-		if err == nil {
-			checkdir := filepath.Join(tmpdir, repo, valcfg.Bidscfg.BidsRoot)
-			if fi, err = os.Stat(checkdir); err == nil && fi.IsDir() {
-				valroot = checkdir
-				log.ShowWrite("[Info] using validation root directory: %s\n%s", valroot, checkdir)
-			} else {
-				log.ShowWrite("[Error] reading validation root directory: %s", err.Error())
-			}
-			validateNifti = valcfg.Bidscfg.ValidateNifti
-		} else {
-			log.ShowWrite("[Error] unmarshalling validation config file: %s", err.Error())
-		}
-	} else {
-		log.ShowWrite("[Info] no validation config file found or processed, running from repo root (%s)", err.Error())
-	}
-
-	// Ignoring NiftiHeaders for now, since it seems to be a common error
-	outBadge := filepath.Join(resdir, srvcfg.Label.ResultsBadge)
-	log.ShowWrite("[Info] Running bids validation: '%s %t --json %s'", srvcfg.Exec.BIDS, validateNifti, valroot)
-
-	// Make sure the validator arguments are in the right order
-	var args []string
-	if !validateNifti {
-		args = append(args, "--ignoreNiftiHeaders")
-	}
-	args = append(args, "--json")
-	args = append(args, valroot)
-
-	// cmd = exec.Command(srvcfg.Exec.BIDS, validateNifti, "--json", valroot)
-	var out, serr bytes.Buffer
-	cmd := exec.Command(srvcfg.Exec.BIDS, args...)
-	out.Reset()
-	serr.Reset()
-	cmd.Stdout = &out
-	cmd.Stderr = &serr
-	cmd.Dir = tmpdir
-	if err = cmd.Run(); err != nil {
-		log.ShowWrite("[Error] running bids validation (%s): '%s', '%s', '%s'",
-			valroot, err.Error(), serr.String(), out.String())
-
-		err = ioutil.WriteFile(outBadge, []byte(resources.BidsFailure), os.ModePerm)
-		if err != nil {
-			log.ShowWrite("[Error] writing results badge for '%s'", repopath)
-		}
-		return 0, err
-	}
-
-	// We need this for both the writing of the result and the badge
-	output := out.Bytes()
-
-	// CHECK: can this lead to a race condition, if a job for the same user/repo combination is started twice in short succession?
-	outFile := filepath.Join(resdir, srvcfg.Label.ResultsFile)
-	err = ioutil.WriteFile(outFile, []byte(output), os.ModePerm)
-	if err != nil {
-		log.ShowWrite("[Error] writing results file for '%s'", repopath)
+	switch validator {
+	case "bids":
+		err = validateBIDS(valroot, resdir)
```

Comment:
We should probably restructure the validation process
- Once a validation has started, the service should create a `processing` badge for the commit and return, so the user does not have to wait until cloning and validation is done (which might take some time).
- Currently a validation process never returns an error. If something goes wrong during the validation an error should be returned here; if an error is present, the service should then replace the `processing` badge with an appropriate `error` badge. If no error is present it is assumed that the build went through and an appropriate badge has been created by the validation process.